### ISSUE TITLE
ControllerLinkBuilder.methodOn not working with @GetMapping

### ISSUE THREAD

**Issue by WinterN** [spring-projects/spring-hateoas#514]
Doing some refactors to adopt new Spring 4.3 request annotations (@GetMapping @PutMapping @PostMapping...) I found that methodOn is not working.

With code is working:

```Java
@RestController
@RequestMapping(value = "/adverts")
@ExposesResourceFor(Advert.class)
public class AdvertController {
  @RequestMapping(method = RequestMethod.GET, value = "/{id}/publications")
  public Resources<Publication> showPublications(@PathVariable("id") Long id) {
    Link selfLink = linkTo(
      methodOn(AdvertController.class).showPublications(id)
    ).withSelfRel();
    //...
  }
}
```
And it returns `/adverts/1/publications` correcty. But if I replace `@RequestMapping` with  `@GetMapping("/{id}/publications")` it only returns `/adverts`.

**Comment by thomasdarimont** [comment]
Just bumped into the same problem.

Did a quick PoC fix in `org.springframework.hateoas.core.AnnotationMappingDiscoverer.getMapping(Class<?>, Method)`

```java
@Override
	public String getMapping(Class<?> type, Method method) {

		Assert.notNull(type, "Type must not be null!");
		Assert.notNull(method, "Method must not be null!");
		
		Annotation annotation = findAnnotation(method, annotationType);
		String[] mapping = getMappingFrom(annotation);
		if (mapping == null || mapping.length == 0) {
			// if no direct mapping could be found look for meta-annotations
			for (Annotation anno : method.getAnnotations()) {
				if (AnnotationUtils.isAnnotationMetaPresent(anno.getClass(), annotationType)){
					mapping = getMappingFrom(anno);
					break;
				}
			}
		}
		
		String typeMapping = getMapping(type);

		if (mapping == null || mapping.length == 0) {
			return typeMapping;
		}

		return typeMapping == null || "/".equals(typeMapping) ? mapping[0] : join(typeMapping, mapping[0]);
	}
```

It works for me with this little patch.
I'm sure @olivergierke knows a better way to solve this ;-)

Cheers,
Thomas

**Comment by olivergierke** [comment]
Duplicates the already solved #471.
